app/src/main/java/org/openobservatory/ooniprobe/domain/QueryDataSource.java

```diff
@@ -33,26 +38,31 @@ public Integer getRefreshKey(@NonNull PagingState<Integer, Result> pagingState)
     public LoadResult.Page<Integer, Result> load(@NonNull LoadParams<Integer> loadParams, @NonNull Continuation<? super LoadResult<Integer, Result>> continuation) {
         // Key may be null during a refresh, if no explicit key is passed into Pager
         // construction. Use 0 as default, because our API is indexed started at index 0
-        int pageNumber = loadParams.getKey() != null ? loadParams.getKey() : 0;
+        int highestId = loadParams.getKey() != null ? loadParams.getKey() : 0;
 
-        SQLOperator[] conditions = (testGroupNameFilter != null && !testGroupNameFilter.isEmpty())
-                ? new SQLOperator[]{Result_Table.test_group_name.is(testGroupNameFilter)}
-                : new SQLOperator[0];
+        List<SQLOperator> conditions = (testGroupNameFilter != null && !testGroupNameFilter.isEmpty())
+                ? Collections.singletonList(Result_Table.test_group_name.is(testGroupNameFilter))
+                : new ArrayList<>();
+        if (highestId != 0) {
+            conditions.add(Result_Table.id.lessThan(highestId));
+        }
 
         List<Result> response = SQLite.select().from(Result.class)
-                .where(conditions)
+                .where(conditions.toArray(new SQLOperator[0]))
                 .limit(20)
-                .offset(pageNumber * 20)
-                .orderBy(Result_Table.start_time, false)
+                .orderBy(Result_Table.id, false)
                 .queryList();
 
-        // Since 0 is the lowest page number, return null to signify no more pages should
-        // be loaded before it.
-        Integer prevKey = (pageNumber > 0) ? pageNumber - 1 : null;
-
+        Integer prevKey;
+        try {
+            prevKey = (!response.isEmpty()) ? Objects.requireNonNull(Iterables.getFirst(response, null)).id - 20 : null;
```

Comment:
Do you actually need the previous key? In which circumstances are you using this?

I don't think you can derive this by just subtracting 20, because some test results may have been removed and so the list of IDs may have "holes" in it.

Reply:
We need the previous key to return it and allow navigation, IIUC the code.

I don't know how the code behaves when there are holes. It may be that, while you are scrolling back, the Android code will continue to query until it has enough elements to fill the screen, so it may be that we don't need to be more precise at the cost of executing more queries.

I think @aanorbel should clarify this.

Reply:
To optimize the load time and memory usage, the records loaded will not always stay in memory, so the previous key is used when the records previously loaded are no longet in memory.